### ISSUE TITLE
UX for client download aborts

### ISSUE THREAD

**Issue by ewindisch** [docker/docker#6928]
Attempting to abort a download usually begins with a user initiating a Ctrl-C against a docker client. This, however, continues the download from the server side and does not have a clear way to resume the download from the perspective of the user.

Suggested solutions are:

1. Detect disconnected clients and abort downloads on the server. Potentially catch ctrl-z or accept a flag to allow pulls to operate in the background without a connected client.
2. Catch ctrl-c and print a warning that the download will continue in the background. Improve reconnecting to 'docker pull' event streams.
3. Implement ' docker pull --list' and 'docker pull --abort', or a set of 'docker download' sub-commands for management of running download tasks.

**Comment by omps** [comment]
IMHO the Ctrl - C should completly stop any of the docker process that is running, something like killalll. there should not be any runaway processes that may be getting in the background for any purpose, let alone be downloading anything.

**Comment by TimWolla** [comment]
:+1:

**Comment by jberkus** [comment]
I understand the arguments around not wanting ctrl-C to kill the running pull because it's difficult to tell the difference between interrupts, and we want different behavior on a large-scale server environment where interrupts may occur for non-user-interactive reasons.  So, I have some suggestions to improve the UX from titaniumrain's plan above.

1. Let's have all the commands titaniumrain suggests.

2. When starting a container in -i interactive mode, and *only* when starting it that way, a CTRL-C should kill the pull, run, or whatever stage the container is at.

3. Add a --no-pull switch for "docker run".  This says "only run this container if it's already pulled, otherwise abort".   This would solve the issue for travelling laptops.

If people think this is a good plan, I can open separate bugs for each of these elements.

**Comment by jberkus** [comment]
Oh, and one more thing:  

4. If a pull received an interrupt (i.e. CTRL-C), then it should print out a warning on STDOUT:

"WARNING: interrupting your pull command does not halt the pull.  Use docker pull --rm to stop the pull if it is still running."

**Comment by pwaller** [comment]
Just to argue the other side, here's what I wrote [over here](https://github.com/docker/docker/pull/9774#issuecomment-72099438):

> Well, I would argue the same applies for pulls, really. In my ideal world, pulls would be reference counted and when the last docker client requesting that pull disconnects the pull should be cancelled. Otherwise again, it might consume expensive resources (e.g, mobile bandwidth), if I set something going and then can't cancel it.

[See also, this post](https://github.com/docker/docker/pull/9774#issuecomment-72091928).

**Comment by bwight** [comment]
In my scenario I get this error when downloading the images using Marathon + Mesos. Marathon is downloading the image and causes a timeout after 5 minutes lets say. I can adjust the timeout on Marathon to 30 minutes if that's what I need to do but its not optimal. If the download takes more than 5 minutes then the image would never be downloaded if aborted when interrupted by the timeout. 

Maybe another way to handle this is to handle pulls in the background and not have to worry about CTRL-C functionality. As soon as I do a pull the command ends quickly.

I liked the commands listed above but think we should add a few more

docker pull --list to list all the current running docker pulls
docker pull --resume to resume a given pull
docker pull --rm to remove a given pull
docker pull --pause to pause a given pull
docker pull --progress shows the progress of a pull and the output that was previously shown

**Comment by fruitl00p** [comment]
:+1: I've noticed a long (_really_ slow) pull of a small image due to an appearant network glitch somewhere down the line. Thus I'd like to cancel and restart. After a ``` CTRL-C ``` i now have lost _all_ insight into the running pull. Running the pull command for the same image results in ```Repository nginx already being pulled by another client. Waiting.``` (v 1.6.0)

Thanks!

**Comment by yebyen** [comment]
@fruitl00p Yep, that's what this issue is for.

Meanwhile, if you don't have any important containers running on that host, or if you don't mind stopping and restarting them, you can restart the docker daemon itself and I believe that will abort your pull.

**Comment by fruitl00p** [comment]
@yebyen Thanks. Already did that in order to by pass the issue. (and upgraded to 1.6.2 while at it) Just wanted to chime in ;)

**Comment by shofetim** [comment]
There really should be some way of killing a push or pull other than restarting the daemon.

The other features would be nice,

- Ability to see the progress after closing the client's terminal
- Potentially binding it to ^C
- See containers that are pending on the pulls completion

But how the functionality is evoked doesn't matter to me nearly as much as having it available.

**Comment by ConnorDoyle** [comment]
+1

**Comment by ruffsl** [comment]
+1

**Comment by jberkus** [comment]
BTW, I've just realized that the situation with "push" is even worse than "pull", because the only way to abort a push is to kill the docker daemon.

**Comment by ezeql** [comment]
+1

**Comment by himulawang** [comment]
+1

**Comment by royingantaginting** [comment]
+1

**Comment by bwhaley** [comment]
+1

**Comment by Slammamonichio** [comment]
+1!!!!

**Comment by andrewfaria** [comment]
+1

**Comment by BradleyA** [comment]
During a pull aborted with CTRL-C. The next docker pull received; Layer already being pulled by another client. Waiting.  Tried again hours later and received the same error.  Fix:   sudo service docker stop; sudo service docker start    cleared the problem for the next pull.  Running Docker version 1.8.1 on Ubuntu 14.04.3 LTS

**Comment by isuftin** [comment]
Will stopping/starting docker-machine work for this?

**Comment by Starefossen** [comment]
Any update on this? Having major problems with stall pulls lately.

**Comment by whiteadam** [comment]
+1 Only been using docker 3 days and I've already experienced this 5 times on 3 different machines.

**Comment by CarlAmbroselli** [comment]
+1

**Comment by nathandunn** [comment]
``docker-machine  kill default``` worked to unstick this on Mac OS X (assuming that ```default``` is the name of your machine).  I opened up a new docker terminal and it worked.    This is day one of me using docker, so I could be missing something.

**Comment by sameerr25** [comment]
Need a resume feature, desperately.

**Comment by xiaoping378** [comment]
Workaround ?  how to resolve it manually,   rmi .. ????

**Comment by xiaoping378** [comment]
resart docker service or restart OS both didn't work.....

**Comment by christianhuening** [comment]
also still getting the error with docker 1.8.2

**Comment by cetex** [comment]
I'm seeing this issue with docker 1.8.2.
Pull of one layer stalls which makes any tasks we try to after this fail miserably, only solution is to kill the daemon which brings down all services running on the host causing noticeable downtime in production.

Imho, if the pull command is killed (ctrl+c, sigkill or something sent by a managing process) docker should without question abort the pull. Giving up and allowing retries if user wants it is better than silently/stubbornly trying to finishing something that may never finish. :)

**Comment by christianhuening** [comment]
+1 needs to be fixed soon

**Comment by buster** [comment]
I have the same problem and it's unclear to me how to fix it.
I rebooted, i reinstalled docker, i removed /var/lib/docker in between, i made sure no images are present, but every time on "docker run swarm create" i get this error...

**Comment by pbdeuchler** [comment]
+1 still experiencing this issue... This should not be labeled as an enhancement but a bug, as SIGINT is not a "pause/resume" signal but a "polite kill"

**Comment by maclof** [comment]
+1 I'm also still experiencing this.

**Comment by coolljt0725** [comment]
#15489 has added showing pull progress in terminal for inflight pull on the latest master branch, so if type ctrl-c to cancel the pull, the pull continue on daemon, and `docker pull` the same image on other terminal will show the progress but not `...already being pulled` message any more.

**Comment by taxilian** [comment]
I don't get it... how is this not considered a critical issue? I had this happen on a production box and now I can't deploy this package; I have to take a *production system* down and restart docker (which all services run under) in order to fix things? I've been trying to research and everyone just sorta seems like "oh, no problem, just *restart the docker service*," like it's not even a big deal.

I can probably manage to restart this box, but this scares me to death! Some of my systems I can't restart except in the middle of the night on weekends and then only with scheduling; how is a bug whose only fix is "restart docker" possibly not considered a huge major problem?

**Comment by christianhuening** [comment]
+1

**Comment by yebyen** [comment]
@taxilian I am not a docker dev, and I do not mean to be combative, but having services like the one you describe is something that scares me to death even more.  You should be able to restart or rebuild any container if you are following 12-factor, and ideally your whole system can suffer an irrecoverable loss of any one machine without sacrificing life, money, or data.  They have high availability databases that are streaming replication and can be fault tolerant in this way.  You can put load balancers in separate availability zones, etc. so that your service stays up even when a whole AZ goes down.

I agree that it is an issue, but I think "the consultant" would tell you that even (especially) 24/7 production deployments should be able to be scaled horizontally, and they should be redundant (so they are able to be rebooted, in case of necessary kernel upgrades), even if those reboots have to be done one node at a time.  Not all software can do that.  I'm dealing with that myself too, right now.

That being said, there was a decent proposal for (at least UX-wise) how to fix this issue up above, and #9774 and others apparently fixed everything about this except the specific facet "can't abort a pull while it's in progress".  So, whoever is affected by this issue should do research and try submitting a pull request, even if not sure how to fix it.

Docker hub is "freemium" so seems likely to be overloaded, at least some percent of the time.  You can run private registries also.  If I had to guess why this is not being classified as a critical issue, it's most likely because nobody paying for support has asked to have it prioritized.

**Comment by pwaller** [comment]
@yebyen, as an external contributor and author of #9774, it is much easier said than done! (The phrase applies to most of your post, but I am referring specifically to the idea that anyone can make a PR to fix the issue).

It's an architectural problem that requires delicate changes in quite a few places and really understands a thorough understanding of what to do. I could do it, but this is nowhere near high enough on my list of priorities and the last few times I have tried to contribute it took quite a while and a lot of energy to shepard the change into master.

**Comment by taxilian** [comment]
@yebyen Of course I have redundancy; that doesn't mean that I want to have to test it live. I'm just saying that any bug which can leave a system like Docker in a state where the only way to fix it is to do a complete restart of the entire system including all unrelated parts is fundamentally a Really Big Problem.

I do appreciate the efforts of people on docker; I just find it baffling that such a critical issue is still around when the early reports of this issue are years old.  I'd be happy to help, but I don't have nearly the experience with the codebase that something this fundamental would require; I don't even care about having a graceful way to fix it, I just wish that there was some way without taking my whole system down to resolve the issue.

**Comment by yebyen** [comment]
@pwaller I was just reading your #15017 and I now see how it's a lot more complicated than I thought.

**Comment by jberkus** [comment]
@taxilian  This is a github project, and Docker is open source.  The purpose of Issues is so that the users can discuss how we can *all* contribute to fixing issues, and for users to supply detailed technical information to devs on our testing.   As such, your demands for immediate attention are liable to go unheard.  If you want more prompt attention, I suggest buying a support contract with one of the several companies who hack Docker code.

In my assessment of the issues, the Pull/Push issues are likely to require a complete overhaul of the pull/push code, just as the devs have been overhauling orchestration, plugins, and networking, possibly including breaking some APIs.  Given that they've been busy with that, I'm not surprised they haven't touched pull/push yet; you're talking having some of the top docker devs spend 6 months working on it.

**Comment by taxilian** [comment]
@jberkus I apologize that my comments were so easily misunderstood; I wasn't trying to demand immediate attention. Honestly it wouldn't help me at this point anyway -- even someone fixed it today it'd likely be at least a month before I could use it on these machines, since we are running the stable coreos and that lags behind production a bit.  Mainly I'm trying to help inject some reality about the severity of the problem into a project which I really respect and appreciate.  This is a serious issue which (to me at least) is a significant concern about the ongoing viability of using Docker in production.

The only response I hoped for was "yeah, we know it's a big deal.  It hasn't been fixed because of .... but we hope to get to it in .... time frame, though that's dependent on other priorities". I guess I need to bend over further backwards to be non-offensive, since I seem to have offended a lot of people who I wasn't really talking to.

That said, and minor irritation aside, I appreciate the explanations; I had not realized that the issue was linked to thinks so deeply buried in complex code.  From an outside view I would not have expected that, and honestly from many of the comments on these threads it looks a lot like people don't really care.

I leave the floor back open for you to all continue criticizing my efforts to inject some clarity into an issue thread.

**Comment by pbdeuchler** [comment]
@jberkus that's a drastic oversimplification, docker makes money through the use and expansion of the docker ecosystem and furthermore has made promises about their code (production ready!) that this issue violates. This isn't a side project put out for use at user's risk, where your caveats would apply. This is a commercial product and for docker to treat such a blatant design flaw (how is telling me that it's a systemic problem in the design supposed to make me feel better? Was it never considered that pulls would be interrupted?) so lackadaisically makes me question why I should be using their products at all. 

Beyond that I'd like to think that, as engineers, we should be able to discuss issues like this frankly and the docker team would appreciate the burden it puts on users of their software. Being dismissive about a pretty big bug in a core feature that many users have experienced (this is one of like 3 issues that's been opened about it) is the wrong way to run an FOSS project no matter what.

**Comment by yebyen** [comment]
@taxilian I don't think you offended anyone, and I'm sorry if you read that in replies.  But I've been subscribed to this issue for about a year now, and I have only seen approximately 100 different people drop in to say "This issue bit me too!  Sucks!  Hope the next guy fixes it!"

It's definitely a problem.  No doubt.  Buying support is definitely a good option if you don't know how to fix it.  Is it a bug that ranks higher than "annoying" or is it really something that is not at all trivial to fix?  Or is it something that can be worked around almost entirely by following several best practices?

It's not a side project, it's a whole company and an ecosystem of companies in fact, but to your point @pbdeuchler it is totally an open source project and "put out for use at user's risk" as you say, according to the license which says in no uncertain terms 13.2 YOU EXPRESSLY UNDERSTAND AND AGREE THAT YOUR USE OF THE DOCKER SERVICES IS AT YOUR SOLE RISK AND THAT THE DOCKER SERVICES ARE PROVIDED "AS IS" AND "AS AVAILABLE."  So I'm not sure what you're trying to say.  Nobody owes anyone else a patch here.

I don't know how to move it forward other than to try to focus the discussion the issue and bring ideas or code on how to resolve it, and unsympathetically "can the crap" about how it's always someone else who should be making this their priority.

Open source projects always need coders.  That is pretty much invariant.  So, tl;dr You got the job!

**Comment by taxilian** [comment]
Totally agreed on your last two paragraphs. I'm actually quite happy to help with things like that, but my reading of earlier comments is that this isn't something that would be easy for someone new to the project to jump into.

Is there someone who would be willing to mentor an ambitious individual and help them figure this out?  If someone can point me in the right general direction and give me an overview of what the problem is then I'd be happy to take a stab at it; I am concerned that if this is a significantly complex issue it may require more background in the codebase than I could aquire in a reasonable time frame.

**Comment by christianhuening** [comment]
As much as it is true that everybody and nobody is responsible for Open Source projects'  bug-fixing, I simply guess that everybody hoped to come here, post the issue and get a "Ahh of course, I got that covered in my PR a week ago. It'll be in 1.8.4, have a nice day and thanks for all the fish" ;-) 
Github and the folks around are awesome and you can get used to that kind of easy getting-your-shit-fixed flow. 

AFAIK: I didn't have that issue for quite a while now, using Docker 1.8.2 and Artifactory for a private repository over here.

**Comment by pbdeuchler** [comment]
@yebyen It would be hard to see the maintainers of, say, the linux kernel or even something smaller like Django/Rails quoting legalese in their license and making the same argument but we all have our own opinions I guess. Nobody is saying these are moral, legal, or fiduciary responsibilities but such responses go against what I interpret to be the core ethos of FOSS.

**Comment by yebyen** [comment]
I mentioned a few comments ago that I'm not a docker core dev or even an external contributor.  I probably shouldn't even be subscribed to this issue, I haven't been affected by it in some time and I am literally infuriated every time another :+1 comes in, asking when it will ever be fixed.  So, I should leave.

You should definitely try and reproduce the issue on the latest release (or even on git head) before you spend too long on any code fix, either architectural or otherwise.  And @pbdeuchler I was responding specifically to your expression that "This isn't a side project put out for use at user's risk" -- the license says it is, unless you have a support contract with some individual or company that says otherwise.  I can't help with a patch!  I'm sorry about that.

Don't read too much into it, I'm just here proposing workarounds, since it is pretty clear that not everyone is affected by this issue, and some have even expressed doubt that it is a real issue.

**Comment by taxilian** [comment]
Just as an FYI, I am still having this issue with a particular docker image (from dockerhub) w/ docker 1.8.2, so it's still there.

**Comment by pbdeuchler** [comment]
Sounds like everyone has their own frustrations here, sorry if I came across as attacking you personally that wasn't my intention at all @yebyen

**Comment by yebyen** [comment]
@pbdeuchler No biggie, we're all friends here

**Comment by jberkus** [comment]
So in any OSS project, there are three basic reasons for a longstanding bug not to get fixed:

1. it's inconsequential
2. it's hard to fix
3. the project is underresourced

If you comb through this thread and the related threads, you will see comments and sometimes patches from several of the major docker contributors, such as @jfrazelle.  Clearly Docker doesn't think this issue is inconsequential, then.  That leaves "hard to fix" or "underresourced", neither of which are liable to be improved by hyperbole in bug reports.   If it's hard to fix, strident complaints don't make it less hard; if the project is underresourced, longer bug threads just make the underresourcing worse.

@pbdeuchler you have clearly NEVER posted a bug to LKML.  Legalese would be a lot kinder than the responses you'd actually get.

@taxilian it's really a matter of tone.  You could have had the same questions, with different wording, and it wouldn't have come across as "fix this now because I need it", which is how your message came across to me and several others.  I've done the same on other projects, and have been told off as well (as I should have been) so I recognize the pattern.

(and again, I'm not a Docker dev)

**Comment by fruitl00p** [comment]
Just to chime in since we ran into this in the course of our docker history: Anyone fearing the 'reboot' daemon solution should really take another look at their stack to see why and try to fix it. This stuff happens, the bug is annoying but its just a minor bug (imho) Everything still works, docker-engine is great life is fine. You just have to either sit out the wait ór restart the daemon.. Either of which shouldn't be the end of the world... Just a minor nuisance. (again: IMHO)

Good luck to the docker team hich i highly respect as they've set various standards (as far as i can tell) on how to do a disruptive innovation like this all out in the open! :+1: to them!

**Comment by buster** [comment]
How someone can say it is a minor bug when the software that i want to
start/run isn't available, is truely beyond me.
All of this OSS/contribution discussion aside, it's NOT a mode of operation
for a serious service to "just keep restarting the server/process until it
eventually works". If it is not clear if a service can be started that is a
major risk for every service. And if that service is redundant and
stateless, that doesn't make the risk go away. It's probably lower, but not
gone.

It is also not true that you can design every service so that you can just
reboot random services without any impact on the service. In the best case
people may lose sessions/connections, in the worst case they just were
about to finish a transaction. It's not very desirable to have either the
option
- to wait for a scheduled downtime of service A because service B doesn't
start on the same docker host
or
- just restart services and maybe cause other problems.

BUT, people who rely on docker for their business and make money with it
should probably pay Docker a buck or two for support. (not sure if docker
devs have said that this is a minor bug, but if so, it is a frightening
stance).

Robin Speekenbrink <notifications@github.com> schrieb am Fr., 16. Okt. 2015
um 08:49 Uhr:

**Comment by cpuguy83** [comment]
Yes, the maintainers are aware of this issue. The push/pull code is... delicate... and is being actively worked on. Is this going to be fixed in 1.9.0? Not likely.

@buster It sounds like the issues you are having are not related to actually aborting a pull.
1.9 includes fixes for these broken pull issues (ie, "a client is already pulling" errors)... including actual bug fixes for the pull code as well re-attaching to the ongoing pull instead of getting that error, fixes for concurrent pulls, etc.

**Comment by coolljt0725** [comment]
@cpuguy83  I'm working on the fix and have made some progress, I have implemented the cancellation of v1_pull on local, maybe I can open a PR several days later....

**Comment by thaJeztah** [comment]
Thanks @coolljt0725, looking forward to your PR

**Comment by huanzhang** [comment]
+1

**Comment by cpuguy83** [comment]
Closing as this is now implemented in #18353 (via pressing ctrl-c while attached to a running pull) and will be in docker 1.10.

**Comment by mcornella** [comment]
That's excellent news! And from what I gathered reading the PR thread, it also improves transfer times using parallelization. Outstanding job @aaronlehmann :tada: :clap:

**Comment by argh** [comment]
I know this issue was closed, but it feels like two issues were conflated, so I'm not sure which is which.

1. the command-line behavior of the docker daemon when someone hits ^C during an operation
2. the integrity of the docker daemon under some sort of transient network condition wherein images are permanently wedged in an "ongoing pull or build" state.

From the PR comments, it appears that what was fixed was only #1, not #2?

(I'm not using the command line, I'm calling the API directly, and responses never return, they're permanently wedged in a "being pulled by another client" state.

Restarting the daemon works, but as other people have said, this isn't a great operational solution in production.  Although our infrastructure is self-healing, it still leads to slow transactions due to having to wait for timeouts, or other transient glitches, all of which lead to bad user behavior.  And I'm uncomfortable with having code in our agents that detects "this docker daemon seems to be flaky, I'll restart it.")

**Comment by thaJeztah** [comment]
@argh reading the top description of this issue, this is about cancelling fro the command-line, or client disconnects. If you think those other issues should be looked into, could you open a new issue? I think that'd work better, and less confusing. :+1:

**Comment by cassiussa** [comment]
@argh If you have already opened a new issue on the subject, as @thaJeztah has suggested, would you mind please posting the link within this thread?  If not, I may open one if I can collect a few more details on how it happened on my systems, or see if I can replicate the issue again.

After just reading this entire thread, it seems to me that most posts were actually about your # 2 item, regardless of whether or not that's what the OP was about.

**Comment by chinadx** [comment]
# docker pull --list
flag provided but not defined: --list
See '/usr/bin/docker-current pull --help'.

-----------
not solved yet...

**Comment by thaJeztah** [comment]
@chinadx the first post shows three _possible_ solutions. Solution number 1 has been implemented; if you press `ctrl-C` during `docker pull`, or the connection with the client is lost, the pull is now aborted. If you think additional changes are needed, please open a _new_ feature request that describes what's needed (and _why_), instead of commenting on an issue that was marked "resolved" over a year ago.

**Comment by casperdcl** [comment]
I don't think `Ctrl+C` works. Tested on: Windows, WSL2, and MacOS.

1. `docker pull some_large_image`
2. <kbd>Ctrl+C</kbd>
3. check network activity (`Task Manager` or `Activity Monitor`), seems daemon is still downloading

Restarting the daemon (overkill) stops the download.


<details><summary>docker version</summary>

```
Client: Docker Engine - Community
 Cloud integration: v1.0.35+desktop.5
 Version:           24.0.6
 API version:       1.43
 Go version:        go1.20.7
 Git commit:        ed223bc
 Built:             Mon Sep  4 12:32:16 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Desktop
 Engine:
  Version:          24.0.6
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.7
  Git commit:       1a79695
  Built:            Mon Sep  4 12:32:16 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.22
  GitCommit:        8165feabfdfe38c65b599c4993d227328c231fca
 runc:
  Version:          1.1.8
  GitCommit:        v1.1.8-0-g82f18fe
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

</details>
